### PR TITLE
Add ROCm+HIP to build system

### PR DESCRIPTION
**Pull request by pbauman**
First of several PRs for HIP support.

This adds options to the build system for enabling ROCm and HIP support. No actual HIP support in the code is included in the PR, this is just the build system options. I've tried to retain the same defaults as the cuda support. The key option is `--with-hip=yes` to turn on HIP support. The environment variable `ROCM_PATH` controls where the build system looks for ROCm, defaulting to `/opt/rocm`.

### CODE REVIEW

**Review by rfalgout — APPROVED** [review]
Looks good to me.  Thanks!

File: src/config/configure.in [hunk 1]
```diff
@@ -1745,20 +1810,53 @@ then
...
+
+dnl *********************************************************************
+dnl * Check for HIP header
+dnl *********************************************************************
+
+dnl If the user has requested to use HIP, we first check the environment
+dnl for ROCM_PATH to point at the ROCm installation. If that is not found,
+dnl then we default to `/opt/rocm`.
+dnl
+dnl TODO: Add an ARG_WITH for rocm so the user can control the ROCm path
```
Comment (liruipeng) [inline src/config/configure.in:1846]: @pbauman I like this idea of "Adding an ARG_WITH", which has been done in another PR for CUDA.

**Review by liruipeng — APPROVED** [review]
Looks good to me. Thanks @pbauman

**Review by ulrikeyang — APPROVED** [review]
This looks good to me.
